File: kata-solutions/gilded-rose-kata/gilded-rose-typescript/3-gilded-rose-refactored/src/gilded-rose.ts
'use strict';

import internal from "stream";

export class Item {
  name: string = "";
  sellIn: number = 0;
  quality: number = 0;

  constructor(name: string, sellIn: number, quality: number) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }

  public incrementQuality() {
    if (this.quality < 50)
      this.quality = this.quality + 1;
  }

  public decrementQuality() {
    if (this.quality > 0)
      this.quality = this.quality - 1;
  }

  public update() {
    this.decrementQuality();
    this.sellIn = this.sellIn - 1;
    this.decrementQuality();
  }

  public toString(): string {
    return "name: " + this.name + ", sellIn: " + this.sellIn + ", quality: " + this.quality
  }
}

export class AgedBrie extends Item {
  constructor(sellIn: number, quality: number) {
    super("Aged Brie", sellIn, quality);
  }

  public update() {
    this.incrementQuality();
    this.sellIn = this.sellIn - 1;
    this.incrementQuality();
  }
}

export class Sulfuras extends Item {
  constructor(sellIn: number, quality: number) {
    super("Sulfuras, Hand of Ragnaros", sellIn, quality);
  }

  public update() {
  }
}

export class BackstagePass extends Item {
  constructor(sellIn: number, quality: number) {
    super("Backstage passes to a TAFKAL80ETC concert", sellIn, quality);
  }

  public update() {
    if (this.quality < 50)
      this.quality = this.quality + 1;
    if (this.sellIn < 11)
      if (this.quality < 50)
        this.quality = this.quality + 1;
    if (this.sellIn < 6)
      if (this.quality < 50)
        this.quality = this.quality + 1;
    
    this.sellIn = this.sellIn - 1;
    
    if (this.sellIn < 0)
      this.quality = 0;
  }
}

export class GildedRose {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  public updateQuality(): Array<Item> {
    this.items.forEach(item => item.update());
    return this.items;
  }
}

File: kata-solutions/gilded-rose-kata/gilded-rose-python/4-gilded-rose-extended/src/gilded_rose.py
class Quality:
    def __init__(self, quality: int):
        self.value = quality

    def increase(self):
        if self.value < 50:
            self.value += 1

    def decrease(self):
        if self.value > 0:
            self.value -= 1


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = Quality(quality)

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality.value)
    
    def update_today(self):
        self.quality.decrease()

    def update_tomorrow(self):
        pass

    def update_quality(self) -> None:
        self.update_today()
        self.sell_in -= 1
        self.update_tomorrow()

class AgedBrie(Item):
    def __init__(self, sell_in, quality):
        super().__init__("Aged Brie", sell_in, quality)

    def update_today(self) -> None:
        self.quality.increase()

    def update_tomorrow(self):
        if self.sell_in < 0:
            self.quality.increase()


class BackstagePassesTafka(Item):
    def __init__(self, sell_in, quality):
        super().__init__("Backstage passes to a TAFKAL80ETC concert", sell_in,
                         quality)

    def update_today(self) -> None:
        self.quality.increase()
        if self.sell_in < 11:
            self.quality.increase()
        if self.sell_in < 6:
            self.quality.increase()

    def update_tomorrow(self):
        if self.sell_in < 0:
            self.quality = Quality(0)


class Sulfuras(Item):
    def __init__(self, sell_in, quality):
        super().__init__("Sulfuras, Hand of Ragnaros", sell_in, quality)

    def update_quality(self) -> None:
        pass

class ConjuredItem(Item):
    def __init__(self, name, sell_in, quality):
        super().__init__(name, sell_in, quality)

    def update_today(self) -> None:
      self.quality.decrease()
      self.quality.decrease()

class GildedRose:
    def __init__(self, items):
        self.items = items

    def update_quality(self) -> None:
        for item in self.items:
            item.update_quality()
